### DISCUSSION TITLE
TSynth maximum MIDI messages per second

### DISCUSSION THREAD

**Discussion by gruedisueli** [ElectroTechnique/TSynth-Teensy4.1#118] [T4.1 code development]
Hi,
I'm wondering if there is some upper limit for processing of MIDI in messages on the TSynth 4.1. I have updated to the latest firmware.
What I'm finding is the TSynth seems to drop MIDI messages when you send data at a very high rate. For example, 1/32nd notes at 240BPM result in many note messages not getting through. I'm aware of MIDI itself being limited to about 3125 bytes per second, which equates to roughly 1000 messages per second, give or take, given different number of bytes required for different message types. If you are sending 240BPM clock messages that equates to roughly 240beats/60sec * 24clockMessages / beat = 96 messages per second for clock stuff. So, still a lot of headroom for other message types such as note on/off or some CC values.

I also note that when used with external MIDI controllers sending CC values, it seems relatively easy to overload the TSynth. I compare this to other devices with MIDI in that do not seem to face the same issues.

The effects I see: Tysnth drops certain notes. Or if you send some CC values from an external controller, the TSynth freezes and blasts a  humming/buzzing sound at 100% sound volume.

I guess I'm wondering if these issues are specific to my own experience, if they are more general to the TSynth, if anyone has some guidelines as to max MIDI bitrate, and perhaps if there are any bugs anyone is aware of that could potentially be fixed in the future.

Many thanks, loving this powerful open-source device and hoping it can continue to be improved in the future!

**Comment by Electrotechnique** [comment]
Yes, I'm not happy with the responsiveness either. It seems to have a lag between receiving MIDI and and the note being heard that can vary slightly.

**Comment by gruedisueli** [comment]
I've begun reviewing the codebase a bit. I do note one thing that seems like a possible cause of exceptions: the function millis() is implicitly cast to a float in multiple places. Millis() returns a type of unsigned long. This is used, for example, when receiving midi clock signals. I suspect this could cause errors or unusual behavior. I may try to revise this on my end and see if it improves things.

**Reply by Electrotechnique** [reply]
There's a few oversights in the code. I recently had a go at knocking integers down to 8 and 16 bit types to free up RAM. I don't like the MIDI responsiveness though and this does need attention.

**Reply by gruedisueli** [reply]
Huh. That's interesting. I've been developing hardware using the same midi library and haven't run into too many responsiveness issues.

**Reply by Electrotechnique** [reply]
I think there's a lot going on and there is a variable lag between note on and the note playing.
